Example on using RandAugment for training image classification models (#408)
keras-team/keras-io: examples/vision/md/randaugment.md

> # RandAugment for Image Classification for Improved Robustness

Reviewer: ```suggestion
# Automated data augmentation with RandAugment for image classification
```

(This way users who don't know RandAugment can click on the example because it's for "data augmentation that does X". Automated data aug. is the main 🍖 , if I understood the paper correctly - ML for ML.)

Author: I explained it above.

**Author:** [Sayak Paul](https://twitter.com/RisingSayak)<br>
**Date created:** 2021/03/13<br>
**Last modified:** 2021/03/13<br>

> **Description:** RandAugment for training an image classification model with improved robustness.

Reviewer: Suggestion: "robustness" can imply many things IMHO, let's be specific and use "translational invariance of convolutional neural networks" (conv nets suffer from this problem as we've seen works, such as Capsule Nets that are trying to address this).

```suggestion
**Description:** Implementing RandAugment to improve translational invariance during image classification.
```

Author: Actually, RandAugment's efficacy is beyond just translational invariance. It also helps to increase robustness to common corruptions and perturbation. Hence the general usage of the word.

> **Description:** RandAugment for training an image classification model with improved robustness.

Reviewer: Ok, let's emphasize `automated data augmentation` (it's Quoc Le's team, so no wonder they're applied ML to ML 😃 I've just scanned the paper.

```suggestion
**Description:** Automated data augmentation with RandAugment for image classification.
```

Author: This might be even more confusing as there's already something called AutoAugment: [AutoAugment: Learning Augmentation Policies from Data](https://arxiv.org/abs/1805.09501)

Reviewer: It's up to you 😄 It was just a suggestion.

I scanned the paper and, like many of Quoc Le team's breakthroughs recently, I think this one was about "AutoML" or something close to meta-learning (ML to ML). And yes, the authors compare RandAugment with AutoAugment (where the capitalization is used because it the name of the method).

I'd at least squeeze `data augmentation` into the title, otherwise it may appear to be vague ("improved robustness").

(e.g. "Data augmentation with RandAugment for image classification" - the title is short, the description is longer)

Reviewer: I'm super happy you made this example - RandAugment is something I should add to my toolkit! 👍

Author: Actually, I even find the "automated" part to be a bit confusing even in RandAugment. Because we are essentially sampling from choices of augmentation transforms from a pre-defined set of transformations with a specific intensity value to apply those transformations. So, there's nothing that automated in this case I think. 

> I'd at least squeeze data augmentation into the title, otherwise it may appear to be vague ("improved robustness").

I think a bit otherwise, though. The name RandAugment itself tells that it has something to do with data augmentation.  On the other hand, there's enough empirical evidence that shows RandAugment improves the robustness of vision models on a general basis. Hence, I don't find "improved robustness" vague.

Reviewer: > The name RandAugment itself tells that it has something to do with data augmentation. 

OK. I'd remember to think about new users who aren't familiar with most of the advanced tricks. To them, AutoAugment or even MultiHeadedAttention would be not as familiar, as to those who play with that code on the weekly basis.

So, personally, I'd highly recommend

```
**Description:** Data augmentation with RandAugment for image classification. 
```

which would explain new users that the example is about `data augmentation` using `RandAugment` for task `image classification`.

But that's just me 😄 

Thanks again for putting this example together! It's awesome.

<img class="k-inline-icon" src="https://colab.research.google.com/img/colab_favicon.ico"/> [**View in Colab**](https://colab.research.google.com/github/keras-team/keras-io/blob/master/examples/vision/ipynb/randaugment.ipynb)  <span class="k-dot">•</span><img class="k-inline-icon" src="https://github.com/favicon.ico"/> [**GitHub source**](https://github.com/keras-team/keras-io/blob/master/examples/vision/randaugment.py)

> Data augmentation is a very useful technique that helps to improve the translational
> invariance of convolutional neural networks (CNN). RandAugment is a stochastic data
> augmentation routine for vision data and was proposed in
> [RandAugment: Practical automated data augmentation with a reduced search space](https://arxiv.org/abs/1909.13719).
> It is composed of strong augmentation
> transforms like color jitters, Gaussian blurs, saturations, etc. along with more
> traditional augmentation transforms such as random crops.

Reviewer: Great paper + example.

Suggestion: refactor to be more specific from the get-go

- "Data augmentation" -> "RandAugment... data augmentation"
- Add `(Cubik et al., 2019)` to add the first author + year of publication

```suggestion
RandAugment is a stochastic data augmentation routine for vision data that helps improve the translational invariance of convolutional neural networks (CNN). It was proposed in
[RandAugment: Practical automated data augmentation with a reduced search space](https://arxiv.org/abs/1909.13719) (Cubik et al., 2019).
It is composed of strong augmentation transforms, including color jitters, Gaussian blurs, saturations, along with more traditional augmentation transforms, such as random crops.
```

    return rand_aug(images=images.numpy())

```

---
## Create TensorFlow `Dataset` objects

Because `RandAugment` can only process NumPy arrays, it
cannot be applied directly as part of the `Dataset` object (which expects TensorFlow
tensors). To make `RandAugment` part of the dataset, we need to wrap it in a
[`tf.py_function`](https://www.tensorflow.org/api_docs/python/tf/py_function).

A `tf.py_function` is a TensorFlow operation (which, like any other TensorFlow operation,
takes TF tensors as arguments and returns TensorFlow tensors) that is capable of running
arbitrary Python code. Naturally, this Python code can only be executed on CPU (whereas
the rest of the TensorFlow graph can be accelerated on GPU), which in some  cases can
cause significant slowdowns -- however, in this case, the `Dataset` pipeline will run
asynchronously together with the model, and doing preprocessing on CPU will remain
performant.

```python
train_ds_rand = (
    tf.data.Dataset.from_tensor_slices((x_train, y_train))
    .shuffle(BATCH_SIZE * 100)
    .batch(BATCH_SIZE)
    .map(
        lambda x, y: (tf.image.resize(x, (IMAGE_SIZE, IMAGE_SIZE)), y),
        num_parallel_calls=AUTO,
    )

>     # The returned output of `tf.py_function` contains an unncessary axis of

Reviewer: Nit: fix typo

```suggestion
    # The returned output of `tf.py_function` contains an unnecessary axis of
```

Total params: 23,585,290
Trainable params: 23,539,850
Non-trainable params: 45,440
_________________________________________________________________
None

```
</div>
We will train this network on two different versions of our dataset:

* One augmented with RandAugment.
* Another one augmented with `simple_aug`.

Since RandAugment is known to enhance the robustness of models to common perturbations
and corruptions, we will also evaluate our models on the CIFAR-10-C dataset, proposed in
[Benchmarking Neural Network Robustness to Common Corruptions and Perturbations](https://arxiv.org/abs/1903.12261)
by Hendrycks et al. The CIFAR-10-C dataset
consists of 19 different image corruptions and perturbations (for example speckle noise,
fog, Gaussian blur, etc.) that too at varying severity levels. For this example we will
be using the following configuration:
[`cifar10_corrupted/saturate_5`](https://www.tensorflow.org/datasets/catalog/cifar10_corrupted#cifar10_corruptedsaturate_5).
The images from this configuration look like so:

![](https://storage.googleapis.com/tfds-data/visualization/fig/cifar10_corrupted-saturate_5-1.0.0.png)

> For the sake of reproducibility, we serialize the initial random weights of our shallow

Reviewer: Suggestion: refactor

```suggestion
In the interest of reproducibility, we serialize the initial random weights of our shallow
```

79/79 [==============================] - 2s 22ms/step - loss: 1.3316 - accuracy: 0.5205
Test accuracy: 52.05%

```
</div>
---
## Load the CIFAR-10-C dataset and evaluate performance

```python
# Load and prepare the CIFAR-10-C dataset
# (If it's not already downloaded, it takes ~10 minutes of time to download)
cifar_10_c = tfds.load("cifar10_corrupted/saturate_5", split="test", as_supervised=True)
cifar_10_c = cifar_10_c.batch(BATCH_SIZE).map(
    lambda x, y: (tf.image.resize(x, (IMAGE_SIZE, IMAGE_SIZE)), y),
    num_parallel_calls=AUTO,
)

# Evaluate `rand_aug_model`
_, test_acc = rand_aug_model.evaluate(cifar_10_c, verbose=0)
print(
    "Accuracy with RandAugment on CIFAR-10-C (saturate_5): {:.2f}%".format(
        test_acc * 100
    )
)

# Evaluate `simple_aug_model`
_, test_acc = simple_aug_model.evaluate(cifar_10_c, verbose=0)
print(
    "Accuracy with simple_aug on CIFAR-10-C (saturate_5): {:.2f}%".format(
        test_acc * 100
    )
)
```

<div class="k-default-codeblock">
```
Accuracy with RandAugment on CIFAR-10-C (saturate_5): 34.14%
Accuracy with simple_aug on CIFAR-10-C (saturate_5): 46.90%

```
</div>

> For the purpose of this example, we trained the models for only a single epoch. In my
> experiments, I found that with RandAugment the model performs way better (76.64%) than
> the model trained with `simple_aug` (64.80%) on the CIFAR-10-C dataset. Additionally, I
> found that RandAugment helped stabilize the training. You can my experimentation notebook
> [here](https://nbviewer.jupyter.org/github/sayakpaul/Keras-Examples-RandAugment/blob/main/RandAugment.ipynb).

Reviewer: Suggestion: 

- refactor: more "we" and/or "you" and less "I" without changing the meaning 🤷 
- missing word/refactor: "You can my experimentation" -> "You can explore this
[notebook](...)"

```suggestion
For the purpose of this example, we trained the models for only a single epoch. On the
CIFAR-10-C dataset, the model with RandAugment can perform better with a higher accuracy
(for example, 76.64% in one experiment) compared with the model trained with `simple_aug` 
(e.g., 64.80%). RandAugment can also help stabilize the training. You can explore this
[notebook](https://nbviewer.jupyter.org/github/sayakpaul/Keras-Examples-RandAugment/blob/main/RandAugment.ipynb) to check some of the results.
```

> As we can see from the notebook, at the expense of increased training time with RandAugment,

Reviewer: Suggestion: for inclusivity/accessibility, let's replace "as we can see"

```suggestion
In the notebook, you may notice that at the expense of increased training time with RandAugment,
```

Author: Not sure how your suggestion is more correlated with inclusivity/accessibility. Do you mind expanding a little bit more?

Reviewer: "we can see" <- not everyone can see. Also, there may be an assumption that the user already noticed it.

Anyway, it's just a suggestion. I may have borrowed this idea from @heyawhite https://heyawhite.com/project/talk-a11y-empathy/

Author: Never thought through it from this angle. Thank you so much. Will definitely keep it in mind. Being an advocate for [compassionate coding](https://compassionatecoding.com/), I think it's definitely something I should be incorporating while writing.

we are able to carve out far better performance on the CIFAR-10-C dataset. You can
experiment on the other corruption and perturbation settings that come with the
run the same CIFAR-10-C dataset and see if RandAugment helps.

You can also experiment with the different values of `n` and `m` in the `RandAugment`
object. In the [original paper](https://arxiv.org/abs/1909.13719), the authors show
the impact of the individual augmentation transforms for a particular task and a range of

> ablation studies. You are welcome to check them out.

Reviewer: ```suggestion
ablation studies.
```

("You are welcome to check them out." may not be necessary, since you already provided the link to that paper at least 2x) :)

> RandAugment has shown great progress in improving the robustness of deep models for
> computer vision as shown in works like [Noisy Student Training](https://arxiv.org/abs/1911.04252) and
> [FixMatch](https://arxiv.org/abs/2001.07685). This makes RandAugment quite a useful
> recipe for training different vision models.

Reviewer: Suggestion: refactor (again, this is just a suggestion :))

```suggestion
RandAugment can improve the robustness of computer vision models, as demonstrated in
[Noisy Student Training](https://arxiv.org/abs/1911.04252) and
[FixMatch](https://arxiv.org/abs/2001.07685). You may find RandAugment a useful part of the
modern toolkit for training computer vision models.
```
